packages/incremental-merkle-tree/src/incremental-merkle-tree.ts

```diff
@@ -30,33 +29,59 @@ export default class IncrementalMerkleTree {
      * @param depth Tree depth.
      * @param zeroValue Zero values for zeroes.
      * @param arity The number of children for each node.
+     * @param leaves The list of initial leaves.
      */
-    constructor(hash: HashFunction, depth: number, zeroValue: Node, arity = 2) {
+    constructor(hash: HashFunction, depth: number, zeroValue: Node, arity = 2, leaves: Node[] = []) {
         checkParameter(hash, "hash", "function")
         checkParameter(depth, "depth", "number")
         checkParameter(zeroValue, "zeroValue", "number", "string", "bigint")
         checkParameter(arity, "arity", "number")
+        checkParameter(leaves, "leaves", "object")
 
         if (depth < 1 || depth > IncrementalMerkleTree.maxDepth) {
             throw new Error("The tree depth must be between 1 and 32")
         }
 
+        if (leaves.length > arity ** depth) {
+            throw new Error(`The tree cannot contain more than ${arity ** depth} leaves`)
+        }
+
         // Initialize the attributes.
         this._hash = hash
         this._depth = depth
         this._zeroes = []
         this._nodes = []
         this._arity = arity
 
-        for (let i = 0; i < depth; i += 1) {
+        for (let level = 0; level < depth; level += 1) {
             this._zeroes.push(zeroValue)
-            this._nodes[i] = []
+            this._nodes[level] = []
             // There must be a zero value for each tree level (except the root).
             zeroValue = hash(Array(this._arity).fill(zeroValue))
         }
 
-        // The default root is the last zero value.
-        this._root = zeroValue
+        this._nodes[depth] = []
+
+        // It initializes the tree with a list of leaves if there are any.
+        if (leaves.length > 0) {
+            this._nodes[0] = leaves
+
+            for (let level = 0; level < depth; level += 1) {
+                for (let index = 0; index < Math.ceil(this._nodes[level].length / arity); index += 1) {
+                    const position = index * arity
+                    const children = []
+
+                    for (let i = position; i < position + arity; i += 1) {
+                        children.push(this._nodes[level][i] || this.zeroes[level])
```

Comment:
There is a bug here if the default leaf is non-zero and the initial leaves array contains a `0`. If `this._nodes[level][i] == 0` then we get `this.zeroes[level]` which could be non-zero. This is because `0` is falsey.

To fix this use the null coalescing operator `??`

Reply:
Oh good catch! Thanks 👍🏽